{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gouin et al.\n",
    "\n",
    "空間トランスクリプトーム\n",
    "\n",
    "Gouin, K.H., Ing, N., Plummer, J.T. et al. An N-Cadherin 2 expressing epithelial cell subpopulation predicts response to surgery, chemotherapy and immunotherapy in bladder cancer. Nat Commun 12, 4906 (2021). https://doi.org/10.1038/s41467-021-25103-7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import squidpy as sq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p Gouin-2021-Nat_Commun-Bladder_cancer/single_cell\n",
    "!mkdir -p Gouin-2021-Nat_Commun-Bladder_cancer/spatial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### snRNA-seq\n",
    "\n",
    "Single-nuclei RNA-seq data is reposited in Gene Expression Omnibus with accession number [GSE169379](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE169379)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-12-26 18:51:06--  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE169nnn/GSE169379/matrix/GSE169379_series_matrix.txt.gz\n",
      "           => 'Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_series_matrix.txt.gz'\n",
      "Resolving ftp.ncbi.nlm.nih.gov... 130.14.250.7, 165.112.9.229\n",
      "Connecting to ftp.ncbi.nlm.nih.gov|130.14.250.7|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /geo/series/GSE169nnn/GSE169379/matrix ... done.\n",
      "==> SIZE GSE169379_series_matrix.txt.gz ... 6404\n",
      "==> PASV ... done.    ==> RETR GSE169379_series_matrix.txt.gz ... done.\n",
      "Length: 6404 (6.3K) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 6,404       --.-K/s   in 0.04s   \n",
      "\n",
      "2021-12-26 18:51:08 (179 KB/s) - 'Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_series_matrix.txt.gz' saved [6404]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE169nnn/GSE169379/matrix/GSE169379_series_matrix.txt.gz -O Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_series_matrix.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_title</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID_REF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM5199001</th>\n",
       "      <td>B1246-GEX: MIBC_rxn1246</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B1246-GEX</td>\n",
       "      <td>MIBC_rxn1246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199003</th>\n",
       "      <td>B1246-HTO: MIBC_rxn1246</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B1246-HTO</td>\n",
       "      <td>MIBC_rxn1246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199004</th>\n",
       "      <td>B1-GEX: MIBC_rxn1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B1-GEX</td>\n",
       "      <td>MIBC_rxn1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199005</th>\n",
       "      <td>B1-HTO: MIBC_rxn1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B1-HTO</td>\n",
       "      <td>MIBC_rxn1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199006</th>\n",
       "      <td>B2-GEX: MIBC_rxn2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B2-GEX</td>\n",
       "      <td>MIBC_rxn2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199007</th>\n",
       "      <td>B2-HTO: MIBC_rxn2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B2-HTO</td>\n",
       "      <td>MIBC_rxn2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199008</th>\n",
       "      <td>B3-GEX: MIBC_rxn3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B3-GEX</td>\n",
       "      <td>MIBC_rxn3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199009</th>\n",
       "      <td>B3-HTO: MIBC_rxn3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B3-HTO</td>\n",
       "      <td>MIBC_rxn3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199011</th>\n",
       "      <td>B4-GEX: MIBC_rxn4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B4-GEX</td>\n",
       "      <td>MIBC_rxn4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199012</th>\n",
       "      <td>B4-HTO: MIBC_rxn4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B4-HTO</td>\n",
       "      <td>MIBC_rxn4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199013</th>\n",
       "      <td>B5A-GEX: MIBC_rxn5A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B5A-GEX</td>\n",
       "      <td>MIBC_rxn5A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199014</th>\n",
       "      <td>B5A-HTO: MIBC_rxn5A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B5A-HTO</td>\n",
       "      <td>MIBC_rxn5A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199015</th>\n",
       "      <td>B5B-GEX: MIBC_rxn5B</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B5B-GEX</td>\n",
       "      <td>MIBC_rxn5B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199016</th>\n",
       "      <td>B5B-HTO: MIBC_rxn5B</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B5B-HTO</td>\n",
       "      <td>MIBC_rxn5B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199017</th>\n",
       "      <td>B6A-GEX: MIBC_rxn6A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B6A-GEX</td>\n",
       "      <td>MIBC_rxn6A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199019</th>\n",
       "      <td>B6A-HTO: MIBC_rxn6A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B6A-HTO</td>\n",
       "      <td>MIBC_rxn6A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199020</th>\n",
       "      <td>B6B-GEX: MIBC_rxn6B</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B6B-GEX</td>\n",
       "      <td>MIBC_rxn6B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199021</th>\n",
       "      <td>B6B-HTO: MIBC_rxn6B</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B6B-HTO</td>\n",
       "      <td>MIBC_rxn6B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199022</th>\n",
       "      <td>B7A-GEX: MIBC_rxn7A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B7A-GEX</td>\n",
       "      <td>MIBC_rxn7A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199023</th>\n",
       "      <td>B7A-HTO: MIBC_rxn7A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B7A-HTO</td>\n",
       "      <td>MIBC_rxn7A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199025</th>\n",
       "      <td>B7B-GEX: MIBC_rxn7B</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B7B-GEX</td>\n",
       "      <td>MIBC_rxn7B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199026</th>\n",
       "      <td>B7B-HTO: MIBC_rxn7B</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B7B-HTO</td>\n",
       "      <td>MIBC_rxn7B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199027</th>\n",
       "      <td>B8A-GEX: MIBC_rxn8A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B8A-GEX</td>\n",
       "      <td>MIBC_rxn8A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199028</th>\n",
       "      <td>B8A-HTO: MIBC_rxn8A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B8A-HTO</td>\n",
       "      <td>MIBC_rxn8A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199029</th>\n",
       "      <td>B8B-GEX: MIBC_rxn8B</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B8B-GEX</td>\n",
       "      <td>MIBC_rxn8B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199030</th>\n",
       "      <td>B8B-HTO: MIBC_rxn8B</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>muscle-invasive bladder cancer (MIBC)</td>\n",
       "      <td>B8B-HTO</td>\n",
       "      <td>MIBC_rxn8B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199031</th>\n",
       "      <td>normalA: normal_sampleA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>normal bladder</td>\n",
       "      <td>normalA</td>\n",
       "      <td>normal_sampleA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199033</th>\n",
       "      <td>normalB: normal_sampleB</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>normal bladder</td>\n",
       "      <td>normalB</td>\n",
       "      <td>normal_sampleB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199034</th>\n",
       "      <td>normalC: normal_sampleC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>normal bladder</td>\n",
       "      <td>normalC</td>\n",
       "      <td>normal_sampleC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5199035</th>\n",
       "      <td>normalD: normal_sampleD</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>normal bladder</td>\n",
       "      <td>normalD</td>\n",
       "      <td>normal_sampleD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       sample_title      organism  \\\n",
       "ID_REF                                              \n",
       "GSM5199001  B1246-GEX: MIBC_rxn1246  Homo sapiens   \n",
       "GSM5199003  B1246-HTO: MIBC_rxn1246  Homo sapiens   \n",
       "GSM5199004        B1-GEX: MIBC_rxn1  Homo sapiens   \n",
       "GSM5199005        B1-HTO: MIBC_rxn1  Homo sapiens   \n",
       "GSM5199006        B2-GEX: MIBC_rxn2  Homo sapiens   \n",
       "GSM5199007        B2-HTO: MIBC_rxn2  Homo sapiens   \n",
       "GSM5199008        B3-GEX: MIBC_rxn3  Homo sapiens   \n",
       "GSM5199009        B3-HTO: MIBC_rxn3  Homo sapiens   \n",
       "GSM5199011        B4-GEX: MIBC_rxn4  Homo sapiens   \n",
       "GSM5199012        B4-HTO: MIBC_rxn4  Homo sapiens   \n",
       "GSM5199013      B5A-GEX: MIBC_rxn5A  Homo sapiens   \n",
       "GSM5199014      B5A-HTO: MIBC_rxn5A  Homo sapiens   \n",
       "GSM5199015      B5B-GEX: MIBC_rxn5B  Homo sapiens   \n",
       "GSM5199016      B5B-HTO: MIBC_rxn5B  Homo sapiens   \n",
       "GSM5199017      B6A-GEX: MIBC_rxn6A  Homo sapiens   \n",
       "GSM5199019      B6A-HTO: MIBC_rxn6A  Homo sapiens   \n",
       "GSM5199020      B6B-GEX: MIBC_rxn6B  Homo sapiens   \n",
       "GSM5199021      B6B-HTO: MIBC_rxn6B  Homo sapiens   \n",
       "GSM5199022      B7A-GEX: MIBC_rxn7A  Homo sapiens   \n",
       "GSM5199023      B7A-HTO: MIBC_rxn7A  Homo sapiens   \n",
       "GSM5199025      B7B-GEX: MIBC_rxn7B  Homo sapiens   \n",
       "GSM5199026      B7B-HTO: MIBC_rxn7B  Homo sapiens   \n",
       "GSM5199027      B8A-GEX: MIBC_rxn8A  Homo sapiens   \n",
       "GSM5199028      B8A-HTO: MIBC_rxn8A  Homo sapiens   \n",
       "GSM5199029      B8B-GEX: MIBC_rxn8B  Homo sapiens   \n",
       "GSM5199030      B8B-HTO: MIBC_rxn8B  Homo sapiens   \n",
       "GSM5199031  normalA: normal_sampleA  Homo sapiens   \n",
       "GSM5199033  normalB: normal_sampleB  Homo sapiens   \n",
       "GSM5199034  normalC: normal_sampleC  Homo sapiens   \n",
       "GSM5199035  normalD: normal_sampleD  Homo sapiens   \n",
       "\n",
       "                                           tissue  sample_id        donor_id  \n",
       "ID_REF                                                                        \n",
       "GSM5199001  muscle-invasive bladder cancer (MIBC)  B1246-GEX    MIBC_rxn1246  \n",
       "GSM5199003  muscle-invasive bladder cancer (MIBC)  B1246-HTO    MIBC_rxn1246  \n",
       "GSM5199004  muscle-invasive bladder cancer (MIBC)     B1-GEX       MIBC_rxn1  \n",
       "GSM5199005  muscle-invasive bladder cancer (MIBC)     B1-HTO       MIBC_rxn1  \n",
       "GSM5199006  muscle-invasive bladder cancer (MIBC)     B2-GEX       MIBC_rxn2  \n",
       "GSM5199007  muscle-invasive bladder cancer (MIBC)     B2-HTO       MIBC_rxn2  \n",
       "GSM5199008  muscle-invasive bladder cancer (MIBC)     B3-GEX       MIBC_rxn3  \n",
       "GSM5199009  muscle-invasive bladder cancer (MIBC)     B3-HTO       MIBC_rxn3  \n",
       "GSM5199011  muscle-invasive bladder cancer (MIBC)     B4-GEX       MIBC_rxn4  \n",
       "GSM5199012  muscle-invasive bladder cancer (MIBC)     B4-HTO       MIBC_rxn4  \n",
       "GSM5199013  muscle-invasive bladder cancer (MIBC)    B5A-GEX      MIBC_rxn5A  \n",
       "GSM5199014  muscle-invasive bladder cancer (MIBC)    B5A-HTO      MIBC_rxn5A  \n",
       "GSM5199015  muscle-invasive bladder cancer (MIBC)    B5B-GEX      MIBC_rxn5B  \n",
       "GSM5199016  muscle-invasive bladder cancer (MIBC)    B5B-HTO      MIBC_rxn5B  \n",
       "GSM5199017  muscle-invasive bladder cancer (MIBC)    B6A-GEX      MIBC_rxn6A  \n",
       "GSM5199019  muscle-invasive bladder cancer (MIBC)    B6A-HTO      MIBC_rxn6A  \n",
       "GSM5199020  muscle-invasive bladder cancer (MIBC)    B6B-GEX      MIBC_rxn6B  \n",
       "GSM5199021  muscle-invasive bladder cancer (MIBC)    B6B-HTO      MIBC_rxn6B  \n",
       "GSM5199022  muscle-invasive bladder cancer (MIBC)    B7A-GEX      MIBC_rxn7A  \n",
       "GSM5199023  muscle-invasive bladder cancer (MIBC)    B7A-HTO      MIBC_rxn7A  \n",
       "GSM5199025  muscle-invasive bladder cancer (MIBC)    B7B-GEX      MIBC_rxn7B  \n",
       "GSM5199026  muscle-invasive bladder cancer (MIBC)    B7B-HTO      MIBC_rxn7B  \n",
       "GSM5199027  muscle-invasive bladder cancer (MIBC)    B8A-GEX      MIBC_rxn8A  \n",
       "GSM5199028  muscle-invasive bladder cancer (MIBC)    B8A-HTO      MIBC_rxn8A  \n",
       "GSM5199029  muscle-invasive bladder cancer (MIBC)    B8B-GEX      MIBC_rxn8B  \n",
       "GSM5199030  muscle-invasive bladder cancer (MIBC)    B8B-HTO      MIBC_rxn8B  \n",
       "GSM5199031                         normal bladder    normalA  normal_sampleA  \n",
       "GSM5199033                         normal bladder    normalB  normal_sampleB  \n",
       "GSM5199034                         normal bladder    normalC  normal_sampleC  \n",
       "GSM5199035                         normal bladder    normalD  normal_sampleD  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_sample = 30\n",
    "with gzip.open(\"Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_series_matrix.txt.gz\", \"rt\") as fi:\n",
    "    meta = []\n",
    "    for i,j in enumerate(fi):\n",
    "        dat = j.strip('\\n!').split(\"\\t\")\n",
    "        if len(dat) != n_sample + 1: continue\n",
    "        dat = [d.strip('\"') for d in dat]\n",
    "        meta.append(dat)\n",
    "meta = pd.DataFrame(meta).set_index(0).T.set_index(\"ID_REF\").iloc[:,[0, 8, 9]]\n",
    "meta.columns = [\"sample_title\", \"organism\", \"tissue\"]\n",
    "meta = meta.assign(sample_id = lambda x: [y.split(\": \")[0] for y in x.sample_title],\n",
    "                   donor_id = lambda x: [y.split(\": \")[1] for y in x.sample_title],\n",
    "                   tissue = lambda x: [y.split(\": \")[1] for y in x.tissue])\n",
    "meta\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-12-26 18:59:04--  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE169nnn/GSE169379/suppl/GSE169379_MIBC_snSeq.h5ad.gz\n",
      "           => 'GSE169379_MIBC_snSeq.h5ad.gz'\n",
      "Resolving ftp.ncbi.nlm.nih.gov... 130.14.250.13, 165.112.9.229\n",
      "Connecting to ftp.ncbi.nlm.nih.gov|130.14.250.13|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /geo/series/GSE169nnn/GSE169379/suppl ... done.\n",
      "==> SIZE GSE169379_MIBC_snSeq.h5ad.gz ... 473158739\n",
      "==> PASV ... done.    ==> RETR GSE169379_MIBC_snSeq.h5ad.gz ... done.\n",
      "Length: 473158739 (451M) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 473,158,739 5.99MB/s   in 85s    \n",
      "\n",
      "2021-12-26 19:00:31 (5.33 MB/s) - 'GSE169379_MIBC_snSeq.h5ad.gz' saved [473158739]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE169nnn/GSE169379/suppl/GSE169379_MIBC_snSeq.h5ad.gz -O Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_MIBC_snSeq.h5ad.gz\n",
    "!gunzip Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_MIBC_snSeq.h5ad.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-12-26 19:16:36--  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE169nnn/GSE169379/suppl/GSE169379_non_tumor_snSeq.h5ad.gz\n",
      "           => 'Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_non_tumor_snSeq.h5ad.gz'\n",
      "Resolving ftp.ncbi.nlm.nih.gov... 130.14.250.13, 165.112.9.229\n",
      "Connecting to ftp.ncbi.nlm.nih.gov|130.14.250.13|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /geo/series/GSE169nnn/GSE169379/suppl ... done.\n",
      "==> SIZE GSE169379_non_tumor_snSeq.h5ad.gz ... 87784694\n",
      "==> PASV ... done.    ==> RETR GSE169379_non_tumor_snSeq.h5ad.gz ... done.\n",
      "Length: 87784694 (84M) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 87,784,694  4.35MB/s   in 18s    \n",
      "\n",
      "2021-12-26 19:16:56 (4.66 MB/s) - 'Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_non_tumor_snSeq.h5ad.gz' saved [87784694]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE169nnn/GSE169379/suppl/GSE169379_non_tumor_snSeq.h5ad.gz -O Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_non_tumor_snSeq.h5ad.gz\n",
    "!gunzip Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_non_tumor_snSeq.h5ad.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_MIBC_snSeq.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_non_tumor_snSeq.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spatial transcriptomics\n",
    "\n",
    "Spatial transcriptome data is reposited in Gene Expression Omnibus with accession number [GSE171351](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE171351)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-12-26 19:23:01--  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE171nnn/GSE171351/suppl/GSE171351_combined_visium.h5ad.gz\n",
      "           => 'Gouin-2021-Nat_Commun-Bladder_cancer/spatial/GSE171351_combined_visium.h5ad.gz'\n",
      "Resolving ftp.ncbi.nlm.nih.gov... 165.112.9.229, 130.14.250.13\n",
      "Connecting to ftp.ncbi.nlm.nih.gov|165.112.9.229|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /geo/series/GSE171nnn/GSE171351/suppl ... done.\n",
      "==> SIZE GSE171351_combined_visium.h5ad.gz ... 48963566\n",
      "==> PASV ... done.    ==> RETR GSE171351_combined_visium.h5ad.gz ... done.\n",
      "Length: 48963566 (47M) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 48,963,566  6.24MB/s   in 9.5s   \n",
      "\n",
      "2021-12-26 19:23:13 (4.92 MB/s) - 'Gouin-2021-Nat_Commun-Bladder_cancer/spatial/GSE171351_combined_visium.h5ad.gz' saved [48963566]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE171nnn/GSE171351/suppl/GSE171351_combined_visium.h5ad.gz -O Gouin-2021-Nat_Commun-Bladder_cancer/spatial/GSE171351_combined_visium.h5ad.gz\n",
    "!gunzip Gouin-2021-Nat_Commun-Bladder_cancer/spatial/GSE171351_combined_visium.h5ad.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 4086 × 33538\n",
       "    obs: 'in_tissue', 'array_row', 'array_col', 'sampleID', 'Patient', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'n_counts'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
       "    uns: 'spatial'\n",
       "    obsm: 'spatial'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(\"Gouin-2021-Nat_Commun-Bladder_cancer/spatial/GSE171351_combined_visium.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>in_tissue</th>\n",
       "      <th>array_row</th>\n",
       "      <th>array_col</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>Patient</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>log1p_total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>n_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCGTTCGTCCAGG-1-0-0-0</th>\n",
       "      <td>1</td>\n",
       "      <td>52</td>\n",
       "      <td>42</td>\n",
       "      <td>A1</td>\n",
       "      <td>Bladder1204</td>\n",
       "      <td>1221</td>\n",
       "      <td>7.108244</td>\n",
       "      <td>1696.0</td>\n",
       "      <td>7.436617</td>\n",
       "      <td>18.396226</td>\n",
       "      <td>26.356132</td>\n",
       "      <td>38.148585</td>\n",
       "      <td>57.488208</td>\n",
       "      <td>48.0</td>\n",
       "      <td>3.891820</td>\n",
       "      <td>2.830189</td>\n",
       "      <td>1696.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAGACGGTTGAT-1-0-0-0</th>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "      <td>79</td>\n",
       "      <td>A1</td>\n",
       "      <td>Bladder1204</td>\n",
       "      <td>3257</td>\n",
       "      <td>8.088869</td>\n",
       "      <td>6578.0</td>\n",
       "      <td>8.791638</td>\n",
       "      <td>16.950441</td>\n",
       "      <td>23.532989</td>\n",
       "      <td>32.335056</td>\n",
       "      <td>47.978109</td>\n",
       "      <td>113.0</td>\n",
       "      <td>4.736198</td>\n",
       "      <td>1.717847</td>\n",
       "      <td>6578.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTGCTGGCTCCAA-1-0-0-0</th>\n",
       "      <td>1</td>\n",
       "      <td>45</td>\n",
       "      <td>67</td>\n",
       "      <td>A1</td>\n",
       "      <td>Bladder1204</td>\n",
       "      <td>3385</td>\n",
       "      <td>8.127405</td>\n",
       "      <td>6858.0</td>\n",
       "      <td>8.833317</td>\n",
       "      <td>19.072616</td>\n",
       "      <td>25.590551</td>\n",
       "      <td>33.712453</td>\n",
       "      <td>48.235637</td>\n",
       "      <td>159.0</td>\n",
       "      <td>5.075174</td>\n",
       "      <td>2.318460</td>\n",
       "      <td>6858.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGCTACGGACCAT-1-0-0-0</th>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>54</td>\n",
       "      <td>A1</td>\n",
       "      <td>Bladder1204</td>\n",
       "      <td>2392</td>\n",
       "      <td>7.780303</td>\n",
       "      <td>4399.0</td>\n",
       "      <td>8.389359</td>\n",
       "      <td>19.686292</td>\n",
       "      <td>27.301659</td>\n",
       "      <td>36.758354</td>\n",
       "      <td>52.716526</td>\n",
       "      <td>125.0</td>\n",
       "      <td>4.836282</td>\n",
       "      <td>2.841555</td>\n",
       "      <td>4399.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGCTCTCGCGCCG-1-0-0-0</th>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>A1</td>\n",
       "      <td>Bladder1204</td>\n",
       "      <td>1963</td>\n",
       "      <td>7.582738</td>\n",
       "      <td>3184.0</td>\n",
       "      <td>8.066208</td>\n",
       "      <td>20.069095</td>\n",
       "      <td>27.386935</td>\n",
       "      <td>36.777638</td>\n",
       "      <td>54.051508</td>\n",
       "      <td>59.0</td>\n",
       "      <td>4.094345</td>\n",
       "      <td>1.853015</td>\n",
       "      <td>3184.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTGCAGCCACGTCA-1-1</th>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>74</td>\n",
       "      <td>D1</td>\n",
       "      <td>Bladder371</td>\n",
       "      <td>1031</td>\n",
       "      <td>6.939254</td>\n",
       "      <td>1596.0</td>\n",
       "      <td>7.375882</td>\n",
       "      <td>26.190476</td>\n",
       "      <td>34.774436</td>\n",
       "      <td>47.305764</td>\n",
       "      <td>66.729323</td>\n",
       "      <td>129.0</td>\n",
       "      <td>4.867535</td>\n",
       "      <td>8.082706</td>\n",
       "      <td>1596.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTGGCCCTGACAGT-1-1</th>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>60</td>\n",
       "      <td>D1</td>\n",
       "      <td>Bladder371</td>\n",
       "      <td>3102</td>\n",
       "      <td>8.040125</td>\n",
       "      <td>7971.0</td>\n",
       "      <td>8.983690</td>\n",
       "      <td>22.682223</td>\n",
       "      <td>32.003513</td>\n",
       "      <td>43.206624</td>\n",
       "      <td>58.650107</td>\n",
       "      <td>305.0</td>\n",
       "      <td>5.723585</td>\n",
       "      <td>3.826370</td>\n",
       "      <td>7971.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTGTTTCCCGAAAG-1-1</th>\n",
       "      <td>1</td>\n",
       "      <td>51</td>\n",
       "      <td>59</td>\n",
       "      <td>D1</td>\n",
       "      <td>Bladder371</td>\n",
       "      <td>1288</td>\n",
       "      <td>7.161622</td>\n",
       "      <td>2108.0</td>\n",
       "      <td>7.653969</td>\n",
       "      <td>23.292220</td>\n",
       "      <td>32.874763</td>\n",
       "      <td>44.165085</td>\n",
       "      <td>62.618596</td>\n",
       "      <td>120.0</td>\n",
       "      <td>4.795791</td>\n",
       "      <td>5.692600</td>\n",
       "      <td>2108.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTAGCAAATTCGA-1-1</th>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>42</td>\n",
       "      <td>D1</td>\n",
       "      <td>Bladder371</td>\n",
       "      <td>2112</td>\n",
       "      <td>7.655864</td>\n",
       "      <td>4357.0</td>\n",
       "      <td>8.379768</td>\n",
       "      <td>22.125316</td>\n",
       "      <td>31.237090</td>\n",
       "      <td>42.299748</td>\n",
       "      <td>59.742942</td>\n",
       "      <td>245.0</td>\n",
       "      <td>5.505332</td>\n",
       "      <td>5.623136</td>\n",
       "      <td>4357.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTCAGTGTGCTAC-1-1</th>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>64</td>\n",
       "      <td>D1</td>\n",
       "      <td>Bladder371</td>\n",
       "      <td>3010</td>\n",
       "      <td>8.010028</td>\n",
       "      <td>7805.0</td>\n",
       "      <td>8.962648</td>\n",
       "      <td>22.152466</td>\n",
       "      <td>31.800128</td>\n",
       "      <td>42.588085</td>\n",
       "      <td>58.436899</td>\n",
       "      <td>211.0</td>\n",
       "      <td>5.356586</td>\n",
       "      <td>2.703395</td>\n",
       "      <td>7805.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4086 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          in_tissue  array_row  array_col sampleID  \\\n",
       "AAACCGTTCGTCCAGG-1-0-0-0          1         52         42       A1   \n",
       "AAACGAGACGGTTGAT-1-0-0-0          1         35         79       A1   \n",
       "AAACTGCTGGCTCCAA-1-0-0-0          1         45         67       A1   \n",
       "AAAGGCTACGGACCAT-1-0-0-0          1         62         54       A1   \n",
       "AAAGGCTCTCGCGCCG-1-0-0-0          1         55         55       A1   \n",
       "...                             ...        ...        ...      ...   \n",
       "TTGTGCAGCCACGTCA-1-1              1         60         74       D1   \n",
       "TTGTGGCCCTGACAGT-1-1              1         18         60       D1   \n",
       "TTGTGTTTCCCGAAAG-1-1              1         51         59       D1   \n",
       "TTGTTAGCAAATTCGA-1-1              1         22         42       D1   \n",
       "TTGTTCAGTGTGCTAC-1-1              1         24         64       D1   \n",
       "\n",
       "                              Patient  n_genes_by_counts  \\\n",
       "AAACCGTTCGTCCAGG-1-0-0-0  Bladder1204               1221   \n",
       "AAACGAGACGGTTGAT-1-0-0-0  Bladder1204               3257   \n",
       "AAACTGCTGGCTCCAA-1-0-0-0  Bladder1204               3385   \n",
       "AAAGGCTACGGACCAT-1-0-0-0  Bladder1204               2392   \n",
       "AAAGGCTCTCGCGCCG-1-0-0-0  Bladder1204               1963   \n",
       "...                               ...                ...   \n",
       "TTGTGCAGCCACGTCA-1-1       Bladder371               1031   \n",
       "TTGTGGCCCTGACAGT-1-1       Bladder371               3102   \n",
       "TTGTGTTTCCCGAAAG-1-1       Bladder371               1288   \n",
       "TTGTTAGCAAATTCGA-1-1       Bladder371               2112   \n",
       "TTGTTCAGTGTGCTAC-1-1       Bladder371               3010   \n",
       "\n",
       "                          log1p_n_genes_by_counts  total_counts  \\\n",
       "AAACCGTTCGTCCAGG-1-0-0-0                 7.108244        1696.0   \n",
       "AAACGAGACGGTTGAT-1-0-0-0                 8.088869        6578.0   \n",
       "AAACTGCTGGCTCCAA-1-0-0-0                 8.127405        6858.0   \n",
       "AAAGGCTACGGACCAT-1-0-0-0                 7.780303        4399.0   \n",
       "AAAGGCTCTCGCGCCG-1-0-0-0                 7.582738        3184.0   \n",
       "...                                           ...           ...   \n",
       "TTGTGCAGCCACGTCA-1-1                     6.939254        1596.0   \n",
       "TTGTGGCCCTGACAGT-1-1                     8.040125        7971.0   \n",
       "TTGTGTTTCCCGAAAG-1-1                     7.161622        2108.0   \n",
       "TTGTTAGCAAATTCGA-1-1                     7.655864        4357.0   \n",
       "TTGTTCAGTGTGCTAC-1-1                     8.010028        7805.0   \n",
       "\n",
       "                          log1p_total_counts  pct_counts_in_top_50_genes  \\\n",
       "AAACCGTTCGTCCAGG-1-0-0-0            7.436617                   18.396226   \n",
       "AAACGAGACGGTTGAT-1-0-0-0            8.791638                   16.950441   \n",
       "AAACTGCTGGCTCCAA-1-0-0-0            8.833317                   19.072616   \n",
       "AAAGGCTACGGACCAT-1-0-0-0            8.389359                   19.686292   \n",
       "AAAGGCTCTCGCGCCG-1-0-0-0            8.066208                   20.069095   \n",
       "...                                      ...                         ...   \n",
       "TTGTGCAGCCACGTCA-1-1                7.375882                   26.190476   \n",
       "TTGTGGCCCTGACAGT-1-1                8.983690                   22.682223   \n",
       "TTGTGTTTCCCGAAAG-1-1                7.653969                   23.292220   \n",
       "TTGTTAGCAAATTCGA-1-1                8.379768                   22.125316   \n",
       "TTGTTCAGTGTGCTAC-1-1                8.962648                   22.152466   \n",
       "\n",
       "                          pct_counts_in_top_100_genes  \\\n",
       "AAACCGTTCGTCCAGG-1-0-0-0                    26.356132   \n",
       "AAACGAGACGGTTGAT-1-0-0-0                    23.532989   \n",
       "AAACTGCTGGCTCCAA-1-0-0-0                    25.590551   \n",
       "AAAGGCTACGGACCAT-1-0-0-0                    27.301659   \n",
       "AAAGGCTCTCGCGCCG-1-0-0-0                    27.386935   \n",
       "...                                               ...   \n",
       "TTGTGCAGCCACGTCA-1-1                        34.774436   \n",
       "TTGTGGCCCTGACAGT-1-1                        32.003513   \n",
       "TTGTGTTTCCCGAAAG-1-1                        32.874763   \n",
       "TTGTTAGCAAATTCGA-1-1                        31.237090   \n",
       "TTGTTCAGTGTGCTAC-1-1                        31.800128   \n",
       "\n",
       "                          pct_counts_in_top_200_genes  \\\n",
       "AAACCGTTCGTCCAGG-1-0-0-0                    38.148585   \n",
       "AAACGAGACGGTTGAT-1-0-0-0                    32.335056   \n",
       "AAACTGCTGGCTCCAA-1-0-0-0                    33.712453   \n",
       "AAAGGCTACGGACCAT-1-0-0-0                    36.758354   \n",
       "AAAGGCTCTCGCGCCG-1-0-0-0                    36.777638   \n",
       "...                                               ...   \n",
       "TTGTGCAGCCACGTCA-1-1                        47.305764   \n",
       "TTGTGGCCCTGACAGT-1-1                        43.206624   \n",
       "TTGTGTTTCCCGAAAG-1-1                        44.165085   \n",
       "TTGTTAGCAAATTCGA-1-1                        42.299748   \n",
       "TTGTTCAGTGTGCTAC-1-1                        42.588085   \n",
       "\n",
       "                          pct_counts_in_top_500_genes  total_counts_mt  \\\n",
       "AAACCGTTCGTCCAGG-1-0-0-0                    57.488208             48.0   \n",
       "AAACGAGACGGTTGAT-1-0-0-0                    47.978109            113.0   \n",
       "AAACTGCTGGCTCCAA-1-0-0-0                    48.235637            159.0   \n",
       "AAAGGCTACGGACCAT-1-0-0-0                    52.716526            125.0   \n",
       "AAAGGCTCTCGCGCCG-1-0-0-0                    54.051508             59.0   \n",
       "...                                               ...              ...   \n",
       "TTGTGCAGCCACGTCA-1-1                        66.729323            129.0   \n",
       "TTGTGGCCCTGACAGT-1-1                        58.650107            305.0   \n",
       "TTGTGTTTCCCGAAAG-1-1                        62.618596            120.0   \n",
       "TTGTTAGCAAATTCGA-1-1                        59.742942            245.0   \n",
       "TTGTTCAGTGTGCTAC-1-1                        58.436899            211.0   \n",
       "\n",
       "                          log1p_total_counts_mt  pct_counts_mt  n_counts  \n",
       "AAACCGTTCGTCCAGG-1-0-0-0               3.891820       2.830189    1696.0  \n",
       "AAACGAGACGGTTGAT-1-0-0-0               4.736198       1.717847    6578.0  \n",
       "AAACTGCTGGCTCCAA-1-0-0-0               5.075174       2.318460    6858.0  \n",
       "AAAGGCTACGGACCAT-1-0-0-0               4.836282       2.841555    4399.0  \n",
       "AAAGGCTCTCGCGCCG-1-0-0-0               4.094345       1.853015    3184.0  \n",
       "...                                         ...            ...       ...  \n",
       "TTGTGCAGCCACGTCA-1-1                   4.867535       8.082706    1596.0  \n",
       "TTGTGGCCCTGACAGT-1-1                   5.723585       3.826370    7971.0  \n",
       "TTGTGTTTCCCGAAAG-1-1                   4.795791       5.692600    2108.0  \n",
       "TTGTTAGCAAATTCGA-1-1                   5.505332       5.623136    4357.0  \n",
       "TTGTTCAGTGTGCTAC-1-1                   5.356586       2.703395    7805.0  \n",
       "\n",
       "[4086 rows x 17 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/maruyamatooru/opt/anaconda3/envs/spatial/lib/python3.8/site-packages/scanpy/preprocessing/_normalization.py:138: UserWarning: Revieved a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n"
     ]
    }
   ],
   "source": [
    "adata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=[\"mt\"], inplace=True)\n",
    "#sc.pp.filter_cells(adata, min_counts=5000)\n",
    "adata = adata[adata.obs.in_tissue != 0, :]\n",
    "sc.pp.normalize_total(adata, inplace=True)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'sc' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-185e867d3997>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mgenes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"in_tissue\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mid\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"A1\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"B1\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"C1\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"D1\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     sc.pl.spatial(adata[adata.obs.sampleID == id, :], \n\u001b[0m\u001b[1;32m      4\u001b[0m                   \u001b[0mimg_key\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"lowres\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m                   \u001b[0mcolor\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgenes\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'sc' is not defined"
     ]
    }
   ],
   "source": [
    "genes = [\"in_tissue\"]\n",
    "for id in [\"A1\", \"B1\", \"C1\", \"D1\"]:\n",
    "    sc.pl.spatial(adata[adata.obs.sampleID == id, :], \n",
    "                  img_key=\"lowres\", \n",
    "                  color=genes,\n",
    "                  alpha=1, \n",
    "                  library_id=id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "11aef0e89ba9b973f30285940aeae5472c259953d734671c591304e9b2e73249"
  },
  "kernelspec": {
   "display_name": "Python 3.8.5 64-bit ('spatial': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
