<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Test Quarto Book - 3&nbsp; Gouin et al.</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./Rcode.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Gouin et al.</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Test Quarto Book</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Curriculum</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Rcode.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">R code</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spatial_transcriptome.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Gouin et al.</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#data-download" id="toc-data-download" class="nav-link active" data-scroll-target="#data-download"><span class="toc-section-number">3.1</span>  Data download</a>
  <ul class="collapse">
  <li><a href="#snrna-seq" id="toc-snrna-seq" class="nav-link" data-scroll-target="#snrna-seq"><span class="toc-section-number">3.1.1</span>  snRNA-seq</a></li>
  <li><a href="#spatial-transcriptomics" id="toc-spatial-transcriptomics" class="nav-link" data-scroll-target="#spatial-transcriptomics"><span class="toc-section-number">3.1.2</span>  Spatial transcriptomics</a></li>
  </ul></li>
  <li><a href="#analysis" id="toc-analysis" class="nav-link" data-scroll-target="#analysis"><span class="toc-section-number">3.2</span>  Analysis</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Gouin et al.</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>空間トランスクリプトーム</p>
<p>Gouin, K.H., Ing, N., Plummer, J.T. et al.&nbsp;An N-Cadherin 2 expressing epithelial cell subpopulation predicts response to surgery, chemotherapy and immunotherapy in bladder cancer. Nat Commun 12, 4906 (2021). https://doi.org/10.1038/s41467-021-25103-7</p>
<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> gzip</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>pd.options.display.max_columns <span class="op">=</span> <span class="va">None</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> squidpy <span class="im">as</span> sq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="data-download" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="data-download"><span class="header-section-number">3.1</span> Data download</h2>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>mkdir <span class="op">-</span>p Gouin<span class="op">-</span><span class="dv">2021</span><span class="op">-</span>Nat_Commun<span class="op">-</span>Bladder_cancer<span class="op">/</span>single_cell</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>mkdir <span class="op">-</span>p Gouin<span class="op">-</span><span class="dv">2021</span><span class="op">-</span>Nat_Commun<span class="op">-</span>Bladder_cancer<span class="op">/</span>spatial</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="snrna-seq" class="level3" data-number="3.1.1">
<h3 data-number="3.1.1" class="anchored" data-anchor-id="snrna-seq"><span class="header-section-number">3.1.1</span> snRNA-seq</h3>
<p>Single-nuclei RNA-seq data is reposited in Gene Expression Omnibus with accession number <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE169379">GSE169379</a></p>
<section id="metadata" class="level4" data-number="3.1.1.1">
<h4 data-number="3.1.1.1" class="anchored" data-anchor-id="metadata"><span class="header-section-number">3.1.1.1</span> Metadata</h4>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>wget ftp:<span class="op">//</span>ftp.ncbi.nlm.nih.gov<span class="op">/</span>geo<span class="op">/</span>series<span class="op">/</span>GSE169nnn<span class="op">/</span>GSE169379<span class="op">/</span>matrix<span class="op">/</span>GSE169379_series_matrix.txt.gz <span class="op">-</span>O Gouin<span class="op">-</span><span class="dv">2021</span><span class="op">-</span>Nat_Commun<span class="op">-</span>Bladder_cancer<span class="op">/</span>single_cell<span class="op">/</span>GSE169379_series_matrix.txt.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--2021-12-26 18:51:06--  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE169nnn/GSE169379/matrix/GSE169379_series_matrix.txt.gz
           =&gt; 'Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_series_matrix.txt.gz'
Resolving ftp.ncbi.nlm.nih.gov... 130.14.250.7, 165.112.9.229
Connecting to ftp.ncbi.nlm.nih.gov|130.14.250.7|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
==&gt; TYPE I ... done.  ==&gt; CWD (1) /geo/series/GSE169nnn/GSE169379/matrix ... done.
==&gt; SIZE GSE169379_series_matrix.txt.gz ... 6404
==&gt; PASV ... done.    ==&gt; RETR GSE169379_series_matrix.txt.gz ... done.
Length: 6404 (6.3K) (unauthoritative)

100%[======================================&gt;] 6,404       --.-K/s   in 0.04s   

2021-12-26 18:51:08 (179 KB/s) - 'Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_series_matrix.txt.gz' saved [6404]
</code></pre>
</div>
</div>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>n_sample <span class="op">=</span> <span class="dv">30</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> gzip.<span class="bu">open</span>(<span class="st">"Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_series_matrix.txt.gz"</span>, <span class="st">"rt"</span>) <span class="im">as</span> fi:</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    meta <span class="op">=</span> []</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i,j <span class="kw">in</span> <span class="bu">enumerate</span>(fi):</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>        dat <span class="op">=</span> j.strip(<span class="st">'</span><span class="ch">\n</span><span class="st">!'</span>).split(<span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(dat) <span class="op">!=</span> n_sample <span class="op">+</span> <span class="dv">1</span>: <span class="cf">continue</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>        dat <span class="op">=</span> [d.strip(<span class="st">'"'</span>) <span class="cf">for</span> d <span class="kw">in</span> dat]</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>        meta.append(dat)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>meta <span class="op">=</span> pd.DataFrame(meta).set_index(<span class="dv">0</span>).T.set_index(<span class="st">"ID_REF"</span>).iloc[:,[<span class="dv">0</span>, <span class="dv">8</span>, <span class="dv">9</span>]]</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>meta.columns <span class="op">=</span> [<span class="st">"sample_title"</span>, <span class="st">"organism"</span>, <span class="st">"tissue"</span>]</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>meta <span class="op">=</span> meta.assign(sample_id <span class="op">=</span> <span class="kw">lambda</span> x: [y.split(<span class="st">": "</span>)[<span class="dv">0</span>] <span class="cf">for</span> y <span class="kw">in</span> x.sample_title],</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>                   donor_id <span class="op">=</span> <span class="kw">lambda</span> x: [y.split(<span class="st">": "</span>)[<span class="dv">1</span>] <span class="cf">for</span> y <span class="kw">in</span> x.sample_title],</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>                   tissue <span class="op">=</span> <span class="kw">lambda</span> x: [y.split(<span class="st">": "</span>)[<span class="dv">1</span>] <span class="cf">for</span> y <span class="kw">in</span> x.tissue])</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>meta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="12">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sample_title</th>
      <th>organism</th>
      <th>tissue</th>
      <th>sample_id</th>
      <th>donor_id</th>
    </tr>
    <tr>
      <th>ID_REF</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GSM5199001</th>
      <td>B1246-GEX: MIBC_rxn1246</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B1246-GEX</td>
      <td>MIBC_rxn1246</td>
    </tr>
    <tr>
      <th>GSM5199003</th>
      <td>B1246-HTO: MIBC_rxn1246</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B1246-HTO</td>
      <td>MIBC_rxn1246</td>
    </tr>
    <tr>
      <th>GSM5199004</th>
      <td>B1-GEX: MIBC_rxn1</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B1-GEX</td>
      <td>MIBC_rxn1</td>
    </tr>
    <tr>
      <th>GSM5199005</th>
      <td>B1-HTO: MIBC_rxn1</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B1-HTO</td>
      <td>MIBC_rxn1</td>
    </tr>
    <tr>
      <th>GSM5199006</th>
      <td>B2-GEX: MIBC_rxn2</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B2-GEX</td>
      <td>MIBC_rxn2</td>
    </tr>
    <tr>
      <th>GSM5199007</th>
      <td>B2-HTO: MIBC_rxn2</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B2-HTO</td>
      <td>MIBC_rxn2</td>
    </tr>
    <tr>
      <th>GSM5199008</th>
      <td>B3-GEX: MIBC_rxn3</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B3-GEX</td>
      <td>MIBC_rxn3</td>
    </tr>
    <tr>
      <th>GSM5199009</th>
      <td>B3-HTO: MIBC_rxn3</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B3-HTO</td>
      <td>MIBC_rxn3</td>
    </tr>
    <tr>
      <th>GSM5199011</th>
      <td>B4-GEX: MIBC_rxn4</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B4-GEX</td>
      <td>MIBC_rxn4</td>
    </tr>
    <tr>
      <th>GSM5199012</th>
      <td>B4-HTO: MIBC_rxn4</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B4-HTO</td>
      <td>MIBC_rxn4</td>
    </tr>
    <tr>
      <th>GSM5199013</th>
      <td>B5A-GEX: MIBC_rxn5A</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B5A-GEX</td>
      <td>MIBC_rxn5A</td>
    </tr>
    <tr>
      <th>GSM5199014</th>
      <td>B5A-HTO: MIBC_rxn5A</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B5A-HTO</td>
      <td>MIBC_rxn5A</td>
    </tr>
    <tr>
      <th>GSM5199015</th>
      <td>B5B-GEX: MIBC_rxn5B</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B5B-GEX</td>
      <td>MIBC_rxn5B</td>
    </tr>
    <tr>
      <th>GSM5199016</th>
      <td>B5B-HTO: MIBC_rxn5B</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B5B-HTO</td>
      <td>MIBC_rxn5B</td>
    </tr>
    <tr>
      <th>GSM5199017</th>
      <td>B6A-GEX: MIBC_rxn6A</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B6A-GEX</td>
      <td>MIBC_rxn6A</td>
    </tr>
    <tr>
      <th>GSM5199019</th>
      <td>B6A-HTO: MIBC_rxn6A</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B6A-HTO</td>
      <td>MIBC_rxn6A</td>
    </tr>
    <tr>
      <th>GSM5199020</th>
      <td>B6B-GEX: MIBC_rxn6B</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B6B-GEX</td>
      <td>MIBC_rxn6B</td>
    </tr>
    <tr>
      <th>GSM5199021</th>
      <td>B6B-HTO: MIBC_rxn6B</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B6B-HTO</td>
      <td>MIBC_rxn6B</td>
    </tr>
    <tr>
      <th>GSM5199022</th>
      <td>B7A-GEX: MIBC_rxn7A</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B7A-GEX</td>
      <td>MIBC_rxn7A</td>
    </tr>
    <tr>
      <th>GSM5199023</th>
      <td>B7A-HTO: MIBC_rxn7A</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B7A-HTO</td>
      <td>MIBC_rxn7A</td>
    </tr>
    <tr>
      <th>GSM5199025</th>
      <td>B7B-GEX: MIBC_rxn7B</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B7B-GEX</td>
      <td>MIBC_rxn7B</td>
    </tr>
    <tr>
      <th>GSM5199026</th>
      <td>B7B-HTO: MIBC_rxn7B</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B7B-HTO</td>
      <td>MIBC_rxn7B</td>
    </tr>
    <tr>
      <th>GSM5199027</th>
      <td>B8A-GEX: MIBC_rxn8A</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B8A-GEX</td>
      <td>MIBC_rxn8A</td>
    </tr>
    <tr>
      <th>GSM5199028</th>
      <td>B8A-HTO: MIBC_rxn8A</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B8A-HTO</td>
      <td>MIBC_rxn8A</td>
    </tr>
    <tr>
      <th>GSM5199029</th>
      <td>B8B-GEX: MIBC_rxn8B</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B8B-GEX</td>
      <td>MIBC_rxn8B</td>
    </tr>
    <tr>
      <th>GSM5199030</th>
      <td>B8B-HTO: MIBC_rxn8B</td>
      <td>Homo sapiens</td>
      <td>muscle-invasive bladder cancer (MIBC)</td>
      <td>B8B-HTO</td>
      <td>MIBC_rxn8B</td>
    </tr>
    <tr>
      <th>GSM5199031</th>
      <td>normalA: normal_sampleA</td>
      <td>Homo sapiens</td>
      <td>normal bladder</td>
      <td>normalA</td>
      <td>normal_sampleA</td>
    </tr>
    <tr>
      <th>GSM5199033</th>
      <td>normalB: normal_sampleB</td>
      <td>Homo sapiens</td>
      <td>normal bladder</td>
      <td>normalB</td>
      <td>normal_sampleB</td>
    </tr>
    <tr>
      <th>GSM5199034</th>
      <td>normalC: normal_sampleC</td>
      <td>Homo sapiens</td>
      <td>normal bladder</td>
      <td>normalC</td>
      <td>normal_sampleC</td>
    </tr>
    <tr>
      <th>GSM5199035</th>
      <td>normalD: normal_sampleD</td>
      <td>Homo sapiens</td>
      <td>normal bladder</td>
      <td>normalD</td>
      <td>normal_sampleD</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="expression" class="level4" data-number="3.1.1.2">
<h4 data-number="3.1.1.2" class="anchored" data-anchor-id="expression"><span class="header-section-number">3.1.1.2</span> Expression</h4>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>wget ftp:<span class="op">//</span>ftp.ncbi.nlm.nih.gov<span class="op">/</span>geo<span class="op">/</span>series<span class="op">/</span>GSE169nnn<span class="op">/</span>GSE169379<span class="op">/</span>suppl<span class="op">/</span>GSE169379_MIBC_snSeq.h5ad.gz <span class="op">-</span>O Gouin<span class="op">-</span><span class="dv">2021</span><span class="op">-</span>Nat_Commun<span class="op">-</span>Bladder_cancer<span class="op">/</span>single_cell<span class="op">/</span>GSE169379_MIBC_snSeq.h5ad.gz</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>gunzip Gouin<span class="op">-</span><span class="dv">2021</span><span class="op">-</span>Nat_Commun<span class="op">-</span>Bladder_cancer<span class="op">/</span>single_cell<span class="op">/</span>GSE169379_MIBC_snSeq.h5ad.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--2021-12-26 18:59:04--  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE169nnn/GSE169379/suppl/GSE169379_MIBC_snSeq.h5ad.gz
           =&gt; 'GSE169379_MIBC_snSeq.h5ad.gz'
Resolving ftp.ncbi.nlm.nih.gov... 130.14.250.13, 165.112.9.229
Connecting to ftp.ncbi.nlm.nih.gov|130.14.250.13|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
==&gt; TYPE I ... done.  ==&gt; CWD (1) /geo/series/GSE169nnn/GSE169379/suppl ... done.
==&gt; SIZE GSE169379_MIBC_snSeq.h5ad.gz ... 473158739
==&gt; PASV ... done.    ==&gt; RETR GSE169379_MIBC_snSeq.h5ad.gz ... done.
Length: 473158739 (451M) (unauthoritative)

100%[======================================&gt;] 473,158,739 5.99MB/s   in 85s    

2021-12-26 19:00:31 (5.33 MB/s) - 'GSE169379_MIBC_snSeq.h5ad.gz' saved [473158739]
</code></pre>
</div>
</div>
<div class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>wget ftp:<span class="op">//</span>ftp.ncbi.nlm.nih.gov<span class="op">/</span>geo<span class="op">/</span>series<span class="op">/</span>GSE169nnn<span class="op">/</span>GSE169379<span class="op">/</span>suppl<span class="op">/</span>GSE169379_non_tumor_snSeq.h5ad.gz <span class="op">-</span>O Gouin<span class="op">-</span><span class="dv">2021</span><span class="op">-</span>Nat_Commun<span class="op">-</span>Bladder_cancer<span class="op">/</span>single_cell<span class="op">/</span>GSE169379_non_tumor_snSeq.h5ad.gz</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>gunzip Gouin<span class="op">-</span><span class="dv">2021</span><span class="op">-</span>Nat_Commun<span class="op">-</span>Bladder_cancer<span class="op">/</span>single_cell<span class="op">/</span>GSE169379_non_tumor_snSeq.h5ad.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--2021-12-26 19:16:36--  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE169nnn/GSE169379/suppl/GSE169379_non_tumor_snSeq.h5ad.gz
           =&gt; 'Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_non_tumor_snSeq.h5ad.gz'
Resolving ftp.ncbi.nlm.nih.gov... 130.14.250.13, 165.112.9.229
Connecting to ftp.ncbi.nlm.nih.gov|130.14.250.13|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
==&gt; TYPE I ... done.  ==&gt; CWD (1) /geo/series/GSE169nnn/GSE169379/suppl ... done.
==&gt; SIZE GSE169379_non_tumor_snSeq.h5ad.gz ... 87784694
==&gt; PASV ... done.    ==&gt; RETR GSE169379_non_tumor_snSeq.h5ad.gz ... done.
Length: 87784694 (84M) (unauthoritative)

100%[======================================&gt;] 87,784,694  4.35MB/s   in 18s    

2021-12-26 19:16:56 (4.66 MB/s) - 'Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_non_tumor_snSeq.h5ad.gz' saved [87784694]
</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_MIBC_snSeq.h5ad"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"Gouin-2021-Nat_Commun-Bladder_cancer/single_cell/GSE169379_non_tumor_snSeq.h5ad"</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="spatial-transcriptomics" class="level3" data-number="3.1.2">
<h3 data-number="3.1.2" class="anchored" data-anchor-id="spatial-transcriptomics"><span class="header-section-number">3.1.2</span> Spatial transcriptomics</h3>
<p>Spatial transcriptome data is reposited in Gene Expression Omnibus with accession number <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE171351">GSE171351</a></p>
<div class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>wget ftp:<span class="op">//</span>ftp.ncbi.nlm.nih.gov<span class="op">/</span>geo<span class="op">/</span>series<span class="op">/</span>GSE171nnn<span class="op">/</span>GSE171351<span class="op">/</span>suppl<span class="op">/</span>GSE171351_combined_visium.h5ad.gz <span class="op">-</span>O Gouin<span class="op">-</span><span class="dv">2021</span><span class="op">-</span>Nat_Commun<span class="op">-</span>Bladder_cancer<span class="op">/</span>spatial<span class="op">/</span>GSE171351_combined_visium.h5ad.gz</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>gunzip Gouin<span class="op">-</span><span class="dv">2021</span><span class="op">-</span>Nat_Commun<span class="op">-</span>Bladder_cancer<span class="op">/</span>spatial<span class="op">/</span>GSE171351_combined_visium.h5ad.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--2021-12-26 19:23:01--  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE171nnn/GSE171351/suppl/GSE171351_combined_visium.h5ad.gz
           =&gt; 'Gouin-2021-Nat_Commun-Bladder_cancer/spatial/GSE171351_combined_visium.h5ad.gz'
Resolving ftp.ncbi.nlm.nih.gov... 165.112.9.229, 130.14.250.13
Connecting to ftp.ncbi.nlm.nih.gov|165.112.9.229|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
==&gt; TYPE I ... done.  ==&gt; CWD (1) /geo/series/GSE171nnn/GSE171351/suppl ... done.
==&gt; SIZE GSE171351_combined_visium.h5ad.gz ... 48963566
==&gt; PASV ... done.    ==&gt; RETR GSE171351_combined_visium.h5ad.gz ... done.
Length: 48963566 (47M) (unauthoritative)

100%[======================================&gt;] 48,963,566  6.24MB/s   in 9.5s   

2021-12-26 19:23:13 (4.92 MB/s) - 'Gouin-2021-Nat_Commun-Bladder_cancer/spatial/GSE171351_combined_visium.h5ad.gz' saved [48963566]
</code></pre>
</div>
</div>
</section>
</section>
<section id="analysis" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="analysis"><span class="header-section-number">3.2</span> Analysis</h2>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"Gouin-2021-Nat_Commun-Bladder_cancer/spatial/GSE171351_combined_visium.h5ad"</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="2">
<pre><code>AnnData object with n_obs × n_vars = 4086 × 33538
    obs: 'in_tissue', 'array_row', 'array_col', 'sampleID', 'Patient', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'n_counts'
    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'
    uns: 'spatial'
    obsm: 'spatial'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>adata.obs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="34">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>in_tissue</th>
      <th>array_row</th>
      <th>array_col</th>
      <th>sampleID</th>
      <th>Patient</th>
      <th>n_genes_by_counts</th>
      <th>log1p_n_genes_by_counts</th>
      <th>total_counts</th>
      <th>log1p_total_counts</th>
      <th>pct_counts_in_top_50_genes</th>
      <th>pct_counts_in_top_100_genes</th>
      <th>pct_counts_in_top_200_genes</th>
      <th>pct_counts_in_top_500_genes</th>
      <th>total_counts_mt</th>
      <th>log1p_total_counts_mt</th>
      <th>pct_counts_mt</th>
      <th>n_counts</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACCGTTCGTCCAGG-1-0-0-0</th>
      <td>1</td>
      <td>52</td>
      <td>42</td>
      <td>A1</td>
      <td>Bladder1204</td>
      <td>1221</td>
      <td>7.108244</td>
      <td>1696.0</td>
      <td>7.436617</td>
      <td>18.396226</td>
      <td>26.356132</td>
      <td>38.148585</td>
      <td>57.488208</td>
      <td>48.0</td>
      <td>3.891820</td>
      <td>2.830189</td>
      <td>1696.0</td>
    </tr>
    <tr>
      <th>AAACGAGACGGTTGAT-1-0-0-0</th>
      <td>1</td>
      <td>35</td>
      <td>79</td>
      <td>A1</td>
      <td>Bladder1204</td>
      <td>3257</td>
      <td>8.088869</td>
      <td>6578.0</td>
      <td>8.791638</td>
      <td>16.950441</td>
      <td>23.532989</td>
      <td>32.335056</td>
      <td>47.978109</td>
      <td>113.0</td>
      <td>4.736198</td>
      <td>1.717847</td>
      <td>6578.0</td>
    </tr>
    <tr>
      <th>AAACTGCTGGCTCCAA-1-0-0-0</th>
      <td>1</td>
      <td>45</td>
      <td>67</td>
      <td>A1</td>
      <td>Bladder1204</td>
      <td>3385</td>
      <td>8.127405</td>
      <td>6858.0</td>
      <td>8.833317</td>
      <td>19.072616</td>
      <td>25.590551</td>
      <td>33.712453</td>
      <td>48.235637</td>
      <td>159.0</td>
      <td>5.075174</td>
      <td>2.318460</td>
      <td>6858.0</td>
    </tr>
    <tr>
      <th>AAAGGCTACGGACCAT-1-0-0-0</th>
      <td>1</td>
      <td>62</td>
      <td>54</td>
      <td>A1</td>
      <td>Bladder1204</td>
      <td>2392</td>
      <td>7.780303</td>
      <td>4399.0</td>
      <td>8.389359</td>
      <td>19.686292</td>
      <td>27.301659</td>
      <td>36.758354</td>
      <td>52.716526</td>
      <td>125.0</td>
      <td>4.836282</td>
      <td>2.841555</td>
      <td>4399.0</td>
    </tr>
    <tr>
      <th>AAAGGCTCTCGCGCCG-1-0-0-0</th>
      <td>1</td>
      <td>55</td>
      <td>55</td>
      <td>A1</td>
      <td>Bladder1204</td>
      <td>1963</td>
      <td>7.582738</td>
      <td>3184.0</td>
      <td>8.066208</td>
      <td>20.069095</td>
      <td>27.386935</td>
      <td>36.777638</td>
      <td>54.051508</td>
      <td>59.0</td>
      <td>4.094345</td>
      <td>1.853015</td>
      <td>3184.0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TTGTGCAGCCACGTCA-1-1</th>
      <td>1</td>
      <td>60</td>
      <td>74</td>
      <td>D1</td>
      <td>Bladder371</td>
      <td>1031</td>
      <td>6.939254</td>
      <td>1596.0</td>
      <td>7.375882</td>
      <td>26.190476</td>
      <td>34.774436</td>
      <td>47.305764</td>
      <td>66.729323</td>
      <td>129.0</td>
      <td>4.867535</td>
      <td>8.082706</td>
      <td>1596.0</td>
    </tr>
    <tr>
      <th>TTGTGGCCCTGACAGT-1-1</th>
      <td>1</td>
      <td>18</td>
      <td>60</td>
      <td>D1</td>
      <td>Bladder371</td>
      <td>3102</td>
      <td>8.040125</td>
      <td>7971.0</td>
      <td>8.983690</td>
      <td>22.682223</td>
      <td>32.003513</td>
      <td>43.206624</td>
      <td>58.650107</td>
      <td>305.0</td>
      <td>5.723585</td>
      <td>3.826370</td>
      <td>7971.0</td>
    </tr>
    <tr>
      <th>TTGTGTTTCCCGAAAG-1-1</th>
      <td>1</td>
      <td>51</td>
      <td>59</td>
      <td>D1</td>
      <td>Bladder371</td>
      <td>1288</td>
      <td>7.161622</td>
      <td>2108.0</td>
      <td>7.653969</td>
      <td>23.292220</td>
      <td>32.874763</td>
      <td>44.165085</td>
      <td>62.618596</td>
      <td>120.0</td>
      <td>4.795791</td>
      <td>5.692600</td>
      <td>2108.0</td>
    </tr>
    <tr>
      <th>TTGTTAGCAAATTCGA-1-1</th>
      <td>1</td>
      <td>22</td>
      <td>42</td>
      <td>D1</td>
      <td>Bladder371</td>
      <td>2112</td>
      <td>7.655864</td>
      <td>4357.0</td>
      <td>8.379768</td>
      <td>22.125316</td>
      <td>31.237090</td>
      <td>42.299748</td>
      <td>59.742942</td>
      <td>245.0</td>
      <td>5.505332</td>
      <td>5.623136</td>
      <td>4357.0</td>
    </tr>
    <tr>
      <th>TTGTTCAGTGTGCTAC-1-1</th>
      <td>1</td>
      <td>24</td>
      <td>64</td>
      <td>D1</td>
      <td>Bladder371</td>
      <td>3010</td>
      <td>8.010028</td>
      <td>7805.0</td>
      <td>8.962648</td>
      <td>22.152466</td>
      <td>31.800128</td>
      <td>42.588085</td>
      <td>58.436899</td>
      <td>211.0</td>
      <td>5.356586</td>
      <td>2.703395</td>
      <td>7805.0</td>
    </tr>
  </tbody>
</table>
<p>4086 rows × 17 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">"mt"</span>] <span class="op">=</span> adata.var_names.<span class="bu">str</span>.startswith(<span class="st">"MT-"</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata, qc_vars<span class="op">=</span>[<span class="st">"mt"</span>], inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co">#sc.pp.filter_cells(adata, min_counts=5000)</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[adata.obs.in_tissue <span class="op">!=</span> <span class="dv">0</span>, :]</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(adata, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/Users/maruyamatooru/opt/anaconda3/envs/spatial/lib/python3.8/site-packages/scanpy/preprocessing/_normalization.py:138: UserWarning: Revieved a view of an AnnData. Making a copy.
  view_to_actual(adata)</code></pre>
</div>
</div>
<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>genes <span class="op">=</span> [<span class="st">"in_tissue"</span>]</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> <span class="bu">id</span> <span class="kw">in</span> [<span class="st">"A1"</span>, <span class="st">"B1"</span>, <span class="st">"C1"</span>, <span class="st">"D1"</span>]:</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(adata[adata.obs.sampleID <span class="op">==</span> <span class="bu">id</span>, :], </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>                  img_key<span class="op">=</span><span class="st">"lowres"</span>, </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>                  color<span class="op">=</span>genes,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>                  alpha<span class="op">=</span><span class="dv">1</span>, </span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>                  library_id<span class="op">=</span><span class="bu">id</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>NameError: name 'sc' is not defined</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./Rcode.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">R code</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->



</body></html>